### ISSUE TITLE
Some issues about DALI dataloader

### ISSUE THREAD

**Issue by xvyaward** [NVIDIA/retinanet-examples#295]
First, I appreciate the great repo and all the hard works.

I have some questions about implements with the DALI dataloader.

1.
It seems that we need `loss_scale` when we use amp, and current code uses different loss_scale option depending on the dali option.
https://github.com/NVIDIA/retinanet-examples/blob/8f69c390a3009cca74a1a876c17703fcfb13c300/odtk/train.py#L41

what does `loss_scale` actually do? And why we have to use dynamic scale only if when we use dali?

2.
I found that some image preprocessing (augmentations) are applied differently b/w `COCOPipeline` in **dali.py** and `CocoDataset` in **data.py**.

For example,  `COCOPipeline` first pads the image and then normalize, while `CocoDataset` first normalizes and then applies padding.
https://github.com/NVIDIA/retinanet-examples/blob/8f69c390a3009cca74a1a876c17703fcfb13c300/odtk/dali.py#L90
https://github.com/NVIDIA/retinanet-examples/blob/8f69c390a3009cca74a1a876c17703fcfb13c300/odtk/data.py#L113-L123

If my understanding is correct, only DALI uses RandomCrop in the case of **training**. Is it normal?

And for **validation**, only `--max-size` option affects resizing for **dali**,
while both `--resize` and `--max-size` options affect resized validation image size in **data.py**
https://github.com/NVIDIA/retinanet-examples/blob/8f69c390a3009cca74a1a876c17703fcfb13c300/odtk/dali.py#L57-L58
https://github.com/NVIDIA/retinanet-examples/blob/8f69c390a3009cca74a1a876c17703fcfb13c300/odtk/data.py#L52-L60

Actually, I found that this affects mAP results, because resized image size in the two options may be different.

So I wonder which preprocessing style of the two options (w/ and w/o dali) is right and if there is no right answer, which one is preferred in research?

Thanks for reading the long question!

**Comment by yashnv** [comment]
1. That loss scale is needed for Apex's loss scaling in during autocast. We saw that using 128.0 loss scale in Pytorch dataloader (data.py) resulted in early infinite loss, so we changed it to be dynamic, which helped in warmup stages.
2. There is no right answer, problems like these are highly dataset specific. Use whatever works best for your dataset, and gives best cross-validation and generalization accuracy.
